test_conformance/api/test_kernel_local_memory_size.cpp

```diff
@@ -0,0 +1,351 @@
+        != 0)
+    {
+        return TEST_FAIL;
+    }
+
+    cl_ulong kernel_local_usage = 0;
+    size_t param_value_size_ret = 0;
+    error = clGetKernelWorkGroupInfo(
+        kernel, deviceID, CL_KERNEL_LOCAL_MEM_SIZE, sizeof(kernel_local_usage),
+        &kernel_local_usage, &param_value_size_ret);
+    test_error(error,
+               "clGetKernelWorkGroupInfo for CL_KERNEL_LOCAL_MEM_SIZE failed");
+
+    test_assert_error(param_value_size_ret == sizeof(cl_ulong),
+                      "param_value_size_ret failed");
+
+    test_assert_error(kernel_local_usage >= 0, "kernel local mem size failed");
+
+    // Check memory needed to execute empty kernel with __local variable
+    if (create_single_kernel_helper(context, &program, &kernel, 1,
+                                    &local_memory_kernel, "local_memory_kernel")
+        != 0)
+    {
+        return TEST_FAIL;
+    }
+
+    kernel_local_usage = 0;
+    param_value_size_ret = 0;
+    error = clGetKernelWorkGroupInfo(
+        kernel, deviceID, CL_KERNEL_LOCAL_MEM_SIZE, sizeof(kernel_local_usage),
+        &kernel_local_usage, &param_value_size_ret);
+    test_error(error,
+               "clGetKernelWorkGroupInfo for CL_KERNEL_LOCAL_MEM_SIZE failed");
+
+    test_assert_error(param_value_size_ret == sizeof(cl_ulong),
+                      "param_value_size_ret failed");
+
+    constexpr size_t size = 10;
+    constexpr size_t memory = size * sizeof(cl_int);
+
+    const size_t global_work_size[] = { size };
+
+    int data[size];
+    for (size_t i = 0; i < size; i++)
+    {
+        data[i] = 0;
+    }
+    clMemWrapper streams[2];
+
+    streams[0] =
+        clCreateBuffer(context, CL_MEM_READ_WRITE, memory, NULL, &error);
+    test_error(error, "Creating test array failed");
+
+    error = clSetKernelArg(kernel, 0, sizeof(streams[0]), &streams[0]);
+    test_error(error, "Unable to set indexed kernel arguments");
+
+    error = clEnqueueNDRangeKernel(queue, kernel, 1, nullptr, global_work_size,
+                                   nullptr, 0, NULL, nullptr);
+    test_error(error, "clEnqueueNDRangeKernel failed.");
+
+    error = clEnqueueReadBuffer(queue, streams[0], CL_TRUE, 0, memory, data, 0,
+                                NULL, NULL);
+    test_error(error, "clEnqueueReadBuffer failed");
+
+    auto local_memory_kernel_verify = [&]() {
+        constexpr size_t size = 10;
+        int testData[size];
+        for (size_t i = 0; i < size; i++)
+        {
+            testData[i] = i * 2;
+            testData[0] += testData[i];
+        }
+        for (size_t i = 0; i < size; i++)
+        {
+            if (data[i] != testData[i]) return false;
+        }
+        return true;
+    };
+    test_assert_error(local_memory_kernel_verify(),
+                      "local_memory_kernel data verificaion failed");
+
+    test_assert_error(kernel_local_usage >= memory,
+                      "kernel local mem size failed");
```

Comment:
FWIW, this check is failing for me for POCL because the queried value is coming back as zero.  I'm not sure if this is a POCL issue or if POCL is being clever and optimizing away the local memory array - maybe @pjaaskel can check?

Reply:
It shouldn't optimize this case away. I opened a [PoCL issue](https://github.com/pocl/pocl/issues/1629). We'll check what's going on.